Afsql threaded dest driver

bazsi opened the pull request:
This branch converts our afsql driver to use the LogThreadedDestDriver framework, instead of its home-grewn implementation (that has originally been the basis of LogThreadedDestDriver).

This should go in after #2315 because without that some of our stats counters would be missing.

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build FAILURE

mitzkia commented:
Internal installer was generated, and tests were Passed.

gaborznagy commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS

furiel commented on modules/afsql/afsql.c:
```diff
@@ -862,53 +821,49 @@ afsql_dd_build_insert_command(AFSqlDestDriver *self, LogMessage *msg, GString *t
-      return TRUE;
+      /* FIXME: I am not sure it makes sense for this behavior, the SQL
+       * server returned an error for our SQL query and the error is not
+       * about a dropped SQL connection, so most probably our SQL has syntax
```
Optional: because the patch results in the same behaviour as the original code.

> not about a dropped SQL connection

How can we conclude that?

I see `afsql_dd_insert` in the new implementation starts with  `afsql_dd_ensure_accessible_database_table`, which I suppose checks the connection but the connection might drop a bit later, before `afsql_dd_run_insert_query` (extremely low probability but possible). Probably this is why we have `dbi_conn_ping` here.

What if we could make the check it faster? db_conn_error can return an integer as error code. We could watch out for DBI_ERROR_NOCONN.

furiel replied:
Overlooked the red and green lines. Deleting my comments, they are not relevand. That's exactly why @bazsi retained the db_conn_ping call, that I though he removed.

gaborznagy commented:
Removing WIP as it was due to waiting for test results

furiel commented on modules/afsql/afsql.c:
```diff
@@ -862,53 +829,49 @@ afsql_dd_build_insert_command(AFSqlDestDriver *self, LogMessage *msg, GString *t
+       * errors or violates some constraints.  Earlier this just dropped and
+       * retried a single message a few times and then progressed.  */
+
+      return WORKER_INSERT_RESULT_SINGLE_ERROR;
```
With @bazsi we concluded that it would worth return WORKER_INSERT_RESULT_ERROR here, that would retry the batch 3 times. The original implementation retried only the last message in the batch, but this change is reasonable.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build SUCCESS

pzoleex commented:
The sql destination group stats counter is missing after this patch:
It should look like this:
destination;d_sql_8798167f89c74ff2a99f7ced012f1f05;;a;processed;639394

but now we have only dst.sql counters.
config:
@version: 7.0
options {
    keep_hostname(yes);
    keep_timestamp(no);
    stats_level(2);
    use_dns(no);
};
source s_network_3400600e69a248308efd0367417265a1 {
    network(ip(0.0.0.0)
 log_fetch_limit(1000)
 log_iw_size(10000)
 max_connections(10)
 port(514));
};
destination d_sql_8798167f89c74ff2a99f7ced012f1f05 {
    sql(columns("datetime varchar(16)", "host varchar(32)", "program  varchar(100)", "pid varchar(8)", "message  varchar(4000)")
 database("test_db")
 flags(explicit-commits)
 flush_lines(1000)
 host("10.140.1.17")
 indexes("host", "program", "pid")
 log_fifo_size(20000)
 password("foo")
 table(t966945653fd9017adad5_testbot)
 type(mssql)
 username("test_user")
 values("$R_DATE", "$HOST", "$PROGRAM", "$PID", "$MSGONLY"));
};
log {
source(s_network_3400600e69a248308efd0367417265a1);
destination(d_sql_8798167f89c74ff2a99f7ced012f1f05);

flags(flow-control);
};

bazsi commented:
thanks for catching this.

On Thu, Sep 27, 2018 at 4:21 PM pzoleex <notifications@github.com> wrote:

> The sql destination group stats counter is missing after this patch:
> It should look like this:
> destination;d_sql_8798167f89c74ff2a99f7ced012f1f05;;a;processed;639394
>
> but now we have only dst.sql counters.
> config:
> @Version <https://github.com/Version>: 7.0
> options {
> keep_hostname(yes);
> keep_timestamp(no);
> stats_level(2);
> use_dns(no);
> };
> source s_network_3400600e69a248308efd0367417265a1 {
> network(ip(0.0.0.0)
> log_fetch_limit(1000)
> log_iw_size(10000)
> max_connections(10)
> port(514));
> };
> destination d_sql_8798167f89c74ff2a99f7ced012f1f05 {
> sql(columns("datetime varchar(16)", "host varchar(32)", "program
> varchar(100)", "pid varchar(8)", "message varchar(4000)")
> database("test_db")
> flags(explicit-commits)
> flush_lines(1000)
> host("10.140.1.17")
> indexes("host", "program", "pid")
> log_fifo_size(20000)
> password("foo")
> table(t966945653fd9017adad5_testbot)
> type(mssql)
> username("test_user")
> values("$R_DATE", "$HOST", "$PROGRAM", "$PID", "$MSGONLY"));
> };
> log {
> source(s_network_3400600e69a248308efd0367417265a1);
> destination(d_sql_8798167f89c74ff2a99f7ced012f1f05);
>
> flags(flow-control);
> };
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2097#issuecomment-425111088>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AArldo62YuYdhSchzCEK7oz3otTOESPwks5ufN7egaJpZM4UdZ6i>
> .
>

bazsi commented:
I've fixed the missing counter issue in #2291, as it is in fact a LogThreadedDestDriver bug, which seems
to have gone unnoticed. I am not sure if I've introduced that with my recent changes, or if it was like
this all along.

SQL got this bug, because it was migrated to use the framework.

The fix itself is this: 9232c01d942b6d197184ca95c59b6b0026343283

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

furiel commented:
From travis:

```
/home/travis/build/balabit/syslog-ng/modules/afsql/afsql.c: In function ‘afsql_dd_init’:
/home/travis/build/balabit/syslog-ng/modules/afsql/afsql.c:1104:3: error: passing argument 1 of ‘log_threaded_dest_driver_start_workers’ from incompatible pointer type [-Werror]
   return log_threaded_dest_driver_start_workers(s);
   ^
In file included from /home/travis/build/balabit/syslog-ng/modules/afsql/afsql.h:27:0,
                 from /home/travis/build/balabit/syslog-ng/modules/afsql/afsql.c:24:
/home/travis/build/balabit/syslog-ng/lib/logthrdestdrv.h:209:10: note: expected ‘struct LogThreadedDestDriver *’ but argument is of type ‘struct LogPipe *’
 gboolean log_threaded_dest_driver_start_workers(LogThreadedDestDriver *self);
          ^
cc1: all warnings being treated as errors
```

bazsi commented:
I have just rebased against latest master, this should resolve the compilation issues.

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

bazsi commented:
@kira-syslogng retest this please;

kira-syslogng commented:
Build FAILURE

bazsi commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS

Ruha0309i commented:
👍👍👍👍

furiel commented:
@bazsi I checked with valgrind, and I found an invalid read:

Config:
```
@version: 3.18

options {
  time-reopen(1);
  stats-level(2);
};

destination d_sql {
    sql(type(sqlite3)
    database("/run/user/1000/testdb.db")
    table("test")
    columns("datetime", "host", "program", "pid", "message")
    values("{$R_DATE}", "${HOST}", "${PROGRAM}", "${PID}", "${MSGONLY}")
    indexes("datetime", "host", "program", "pid", "message")
    flush-lines(2)
    flags(explicit-commits)
    );
};

log {
    source { stdin(flags(no-parse)); };
    destination(d_sql);
    flags(flow-control);
};
```

The invalid read happens right after syslog-ng starts, without messages.
```
==8475== Invalid read of size 1
==8475==    at 0x94BC3FB: ??? (in /usr/lib/x86_64-linux-gnu/dbd/libdbdsqlite3.so)
==8475==    by 0x799A85E: dbi_conn_connect (in /usr/lib/x86_64-linux-gnu/libdbi.so.1.1.0)
==8475==    by 0x7789B85: afsql_dd_connect (afsql.c:691)
==8475==    by 0x4E873D1: _compat_connect (logthrdestdrv.c:744)
==8475==    by 0x4E85A79: log_threaded_dest_worker_connect (logthrdestdrv.h:153)
==8475==    by 0x4E85FE9: _connect (logthrdestdrv.c:177)
==8475==    by 0x4E86967: _perform_work (logthrdestdrv.c:440)
==8475==    by 0x603C056: ??? (in /usr/lib/x86_64-linux-gnu/libivykis.so.0.5.4)
==8475==    by 0x603F38D: iv_main (in /usr/lib/x86_64-linux-gnu/libivykis.so.0.5.4)
==8475==    by 0x4E87051: _worker_thread (logthrdestdrv.c:640)
==8475==    by 0x4E8E1FE: _worker_thread_func (mainloop-worker.c:339)
==8475==    by 0x51A5C54: ??? (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4800.2)
==8475==  Address 0x7339e4f is 1 bytes before a block of size 26 alloc'd
==8475==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==8475==    by 0x94BC394: ??? (in /usr/lib/x86_64-linux-gnu/dbd/libdbdsqlite3.so)
==8475==    by 0x799A85E: dbi_conn_connect (in /usr/lib/x86_64-linux-gnu/libdbi.so.1.1.0)
==8475==    by 0x7789B85: afsql_dd_connect (afsql.c:691)
==8475==    by 0x4E873D1: _compat_connect (logthrdestdrv.c:744)
==8475==    by 0x4E85A79: log_threaded_dest_worker_connect (logthrdestdrv.h:153)
==8475==    by 0x4E85FE9: _connect (logthrdestdrv.c:177)
==8475==    by 0x4E86967: _perform_work (logthrdestdrv.c:440)
==8475==    by 0x603C056: ??? (in /usr/lib/x86_64-linux-gnu/libivykis.so.0.5.4)
==8475==    by 0x603F38D: iv_main (in /usr/lib/x86_64-linux-gnu/libivykis.so.0.5.4)
==8475==    by 0x4E87051: _worker_thread (logthrdestdrv.c:640)
==8475==    by 0x4E8E1FE: _worker_thread_func (mainloop-worker.c:339)
==8475== 
```

bazsi commented:
This error is in dbd_sqlite3 and is caused by the fact that we set sqlite3_dbdir to an empty string, which that code does not handle properly.

I am thinking about a possible workaround, however this is clearly not the result of this refactor, but has been in the code for quite some time now.

bazsi commented:
I have added a workaround on top of the existing branch.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

furiel commented:
I found another leak, but as the previous one, it is not caused by the refactor. _is_table_present receives a db_res parameter, that is only freed in the true branch. As a result: we leak a leak a dbi_result once, when the table is created. We can handle this independently from this patch. Approve from my side.

pzoleex commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS
